### PR TITLE
get_nonce and get_storage_at; Old state support

### PR DESCRIPTION
**Pull request by marioiordanov**
based on https://github.com/0xSpaceShard/starknet-devnet-rs/pull/103
Most of the files that are changed are coming from the PR that this one is based on

## Usage related changes

RPCs:

- get_nonce
- get_storage_at

## Development related changes

Refactored:

- state's get_storage to not take mutable reference to state
- generate_new_block to accept a Starknet state
- get_state_at doesn't return NotImplementedError

Added:

- mapping of block number to starknet state, so every request can be executed on previous blocks
- new error type NoStateAtBlock in starknet-core

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [ ] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet-server/src/api/json_rpc/endpoints.rs [hunk 1]
```diff
@@ -149,10 +162,14 @@ impl JsonRpcHandler {
         let starknet = self.api.starknet.read().await;
         match starknet.get_class_hash_at(&block_id.into(), &contract_address.0) {
             Ok(class_hash) => Ok(FeltHex(class_hash)),
-            Err(Error::BlockIdHashUnimplementedError | Error::BlockIdNumberUnimplementedError) => {
-                Err(ApiError::BlockNotFound)
+            Err(
+                Error::BlockIdHashUnimplementedError
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/endpoints.rs:166]: Can this still be an error? Can the error definitions be removed?

File: crates/starknet/src/starknet/mod.rs [hunk 2]
```diff
@@ -580,19 +621,24 @@ mod tests {
     #[test]
     fn getting_state_reader_at_block_by_hash() {
```
Comment (FabijanC) [inline crates/starknet/src/starknet/mod.rs:622]: I see I forgot to rename the tests when I renamed the function from `get_state_reader_at` to `get_state_at`. Also, the name of the test could reflect the fact that it's an unhappy path (or that it should fail): `getting_state_reader_at_block_by_nonexistent_hash`

File: crates/starknet/src/starknet/mod.rs [hunk 3]
```diff
@@ -580,19 +621,24 @@ mod tests {
...
+        starknet.generate_new_block(StateDiff::default(), starknet.state.clone()).unwrap();
+
+        match starknet.get_state_at(&BlockId::Hash(Felt::from(0).into())) {
+            Err(Error::NoBlock) => (),
             _ => panic!("Should have failed"),
         }
     }
 
     #[test]
     fn getting_state_reader_at_block_by_number() {
```
Comment (FabijanC) [inline crates/starknet/src/starknet/mod.rs:634]: Same as for `getting_state_reader_at_block_by_hash`

**Review by FabijanC — CHANGES_REQUESTED** [review]
More thorough testing of old state support delegated to issue #108

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs:166]: also on the call method on L208 in endpoints.rs

File: crates/starknet/src/starknet/mod.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs]: [8757a47](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/104/commits/8757a47ecd6e798fc5c85f4251a68aa8734087c0)

File: crates/starknet/src/starknet/mod.rs (hunk 3 above)
Comment (marioiordanov) [reply crates/starknet/src/starknet/mod.rs]: [8757a47](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/104/commits/8757a47ecd6e798fc5c85f4251a68aa8734087c0)

File: crates/starknet-server/src/api/json_rpc/endpoints.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/endpoints.rs]: [3074c87](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/104/commits/3074c877e437aa6c811442fe2bf8be2b33ef7124)

**Review by FabijanC — APPROVED** [review]
Approving. Only conflict resolution remains.

**Pull request merged**